### PR TITLE
DAOS-9196: create pool with all remaining capacity

### PR DESCRIPTION
**Pull request by knard-intel**
Add a new option to dmg pool create to allocate a pool with all the remaining space available on pmem or NVMe (if configured)

This patch has several limitations which could be alleviated with some minor modifications of the communication protocol.  More details could be found in the comments.

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@intel.com>

### CODE REVIEW

File: src/control/lib/control/mocks.go
Comment (tanabarr) [inline src/control/lib/control/mocks.go:565]: I imagine this is probably duplicated somewhere else in the package

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:144]: - [ ] Replace if statement with swith statement

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:264]: From what I understand this multiple hostset should never happen. However, I would like to keep this test as HostStorage response could be agregated in the Future. Moreover, Multiple host should also be properly managed as we are just looking for minimal storage values. Thus, what I propose is to remove the panic with a warning message.
Could tell me, if this solution is acceptable for you ?

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:272]: From my side, there is no issue to remove this sanity test if you think that it is useless (Moreover, I was not able to test it properly). However, could you tell me if there is the same kind of sentence as the C/C++ assert in GO language: from my side, I was not able to find something equivalent.

- [ ] Remove useless RebootRequired sanity check

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:287]: - [ ] Fix spelling eacch -> each

File: src/control/lib/control/mocks.go
Comment (knard-intel) [reply src/control/lib/control/mocks.go:565]: I have already searched for something like this, but I was sadly not able to find it. I will double check it as I am starting to be a little bit more familiar with the code.

- [ ] Look for duplicated code
- [ ] Remove duplicated code if needed

File: src/control/cmd/dmg/pool.go
Comment (tanabarr) [reply src/control/cmd/dmg/pool.go:264]: Yes, you are unlikely to get aggregated results (hostset representing more than one host) with `Usage: true` set in the request with the current server side implementation. If your implementation relies upon responses not being aggregated then you should return an error in the case that aggregated responses are received so that when other areas of implementation change it triggers errors and the functional regressions are captured.

File: src/control/cmd/dmg/pool.go
Comment (tanabarr) [reply src/control/cmd/dmg/pool.go:272]: I think that explicit tests and returned errors are preferred over asserts in Go.

**Review by tanabarr — APPROVED** [review]
there is nothing blocking from my side, just suggestions for improvements if you need to re-push or in a follow-up PR

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:264]: At the end, as my implementation is independent if the host are aggregated or not,  I have totally remove this useless code. I also updated the test to check that agregated hostset are properly handle.

File: src/control/lib/control/mocks.go
Comment (knard-intel) [reply src/control/lib/control/mocks.go:565]: There is indeed a `MockServerScanResp` which is also mocking a `ctlpb.StorageScanResp`. However, I choosed to not use it: using it will imply some heavy modifications (and a noticeable increase of its complexity) of this mocker and a very small portions of code could be reused. Thus, I priviligied to duplicate a small portion of code and keep the thing simple: [KISS ](https://fr.wikipedia.org/wiki/Principe_KISS) ;-)

**Review by tanabarr — APPROVED** [review]

**Review by daltonbohning — COMMENTED** [review]
I'll let someone else review, as I'm not familiar with the GO or control code

File: src/control/cmd/dmg/pool.go
Comment (mjmac) [inline src/control/cmd/dmg/pool.go:74]: Huh. This is clever, haven't seen this construct before but it makes sense. I suggest renaming `errIncompatFlagsNb` to something more like `seenSizeFlags` to make the logic clearer.

Only other suggestion I might make here is that you might want to define the input test array separately with each test on a new line. Slightly more verbose, but will make adding new tests clear.

File: src/control/cmd/dmg/pool.go
Comment (mjmac) [inline src/control/cmd/dmg/pool.go:83]: You could also switch on the number of flags seen; might be slightly easier to read.

File: src/control/cmd/dmg/pool.go
Comment (mjmac) [inline src/control/cmd/dmg/pool.go:242]: Per Go conventions, comments on exported symbols should begin with the name of the entity, e.g. `// PoolMetadataBytes defines the amount of storage reserved for pool metadata.`. This allows for automatically-generated documentation. If the symbol is not relevant outside of the package, then it should not be exported (i.e. the first letter should be lowercase).

File: src/control/cmd/dmg/pool.go
Comment (mjmac) [inline src/control/cmd/dmg/pool.go:245]: This is not the correct place for this logic, because it means that only the `dmg` tool will be able to create auto-sized pools. Instead, this should be built into `lib/control/pool.go` so that Control API users may benefit from it.

**Review by mjmac — CHANGES_REQUESTED** [review]
I have a few concerns about the approach:

1. The biggest concern I have is that there is too much "business logic" in `cmd/dmg` ... Think of dmg as one of several possible UI layers, all of which use the Control API in `lib/control`. Please refactor so that callers of `control.PoolCreate()` may benefit.
2. The ticket does indeed request a new flag for `-a,--all` and this PR implements that request as written. However, as @daltonbohning has pointed out, it would be more flexible and user-friendly to allow the `--size` parameter to accept a percentage value to indicate that the admin wants the pool to use some percentage of available storage.
3. I am a little concerned that querying client-side like this will have some scalability issues. I think I'm OK with taking an iterative approach where the implementation details are hidden behind the `control.PoolCreate()` API, though. As long as we get that interface correct now, we can optimize later without breaking anything.

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:74]: Following tasks will be done:
- [ ] Refactoring `errIncompatFlagsNb` to `seenSizeFlags`
- [ ] Refactoring declare testing array (improvement of code readability and maintenance)

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:83]: Following tasks will be done:
- [ ] Refactoring sequential `i` with `switch`

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:242]: Following tasks will be done:
- [ ] Fix code documentation to respect go idiomatic coding convention

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:245]: Following tasks will be done:
- [ ] Code refactoring, move `GetMaxPoolSize` function to `lib/control/pool.go`

**Comment by knard-intel** [comment]
> I have a few concerns about the approach:
> 
> 1. The biggest concern I have is that there is too much "business logic" in `cmd/dmg` ... Think of dmg as one of several possible UI layers, all of which use the Control API in `lib/control`. Please refactor so that callers of `control.PoolCreate()` may benefit.

Following tasks will be done:
- [ ] Code refactoring, move computational logic to `lib/control/pool.go` to improve code factorization and reusability

> 2. The ticket does indeed request a new flag for `-a,--all` and this PR implements that request as written. However, as @daltonbohning has pointed out, it would be more flexible and user-friendly to allow the `--size` parameter to accept a percentage value to indicate that the admin wants the pool to use some percentage of available storage.

Following tasks will be done:
- [ ]  Manage percentage value

> 3. I am a little concerned that querying client-side like this will have some scalability issues. I think I'm OK with taking an iterative approach where the implementation details are hidden behind the `control.PoolCreate()` API, though. As long as we get that interface correct now, we can optimize later without breaking anything.

I am totally agree that this approach is suboptimal and is just a first step toward an approach without scaling issue. For this latest solution, we will probably have to change the communication protocol to properly manage SCM ranks and reduce the traffic.

**Review by kjacque — COMMENTED** [review]
Generally good, I have a mix of comments and suggestions below.

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go]: Update the behavior of the all flags (changed to all-ratio).
If all-ratio is set => ratio is equal to 100.
all-ratio=XX% => ratio is equal to XX%

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go]: The different size will be printed with the response of the create pool message.

File: src/control/cmd/dmg/pool_test.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool_test.go]: Following tasks will be done:
- [ ] Respect coding style

File: src/control/cmd/dmg/pool_test.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool_test.go:931]: Following tasks will be done:
- [ ] Respect coding style

File: src/control/cmd/dmg/pool_test.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool_test.go]: Following tasks will be done:
- [ ] Remove useless error informative message

File: src/control/cmd/dmg/pool_test.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool_test.go]: Following tasks will be done:

- [ ] Remove useless error informative message

File: src/control/lib/control/pool.go
Comment (knard-intel) [reply src/control/lib/control/pool.go:982]: From my side I would prefer to keep as this. It is an habit from me to manage integer overflow and casting unexpected behavior. However, I am not sure if this remarks is relevant with golang (just started to use it)  ? If my argument is not relevant with golang, I will change it.

File: src/control/lib/control/pool.go
Comment (knard-intel) [reply src/control/lib/control/pool.go:996]: Following tasks will be done:

- [ ] Use formated error

File: src/control/lib/control/pool.go
Comment (knard-intel) [reply src/control/lib/control/pool.go:999]: I put these remarks as reminder and also for creating next tickets.
Currently, I do not know the process to create this kind of follow up tickets.
From my past experience, only the Product Owner is allowed to update the backlog.

File: src/control/lib/control/pool.go
Comment (knard-intel) [reply src/control/lib/control/pool.go]: Sorry, I do not find the spelling or language issue.

File: src/control/lib/control/pool_test.go
Comment (knard-intel) [reply src/control/lib/control/pool_test.go:1358]: Following tasks will be done:

- [ ] Factorize code or split function

File: src/control/cmd/dmg/pool_test.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool_test.go]: Following tasks will be done:

- [ ] Investigate if cmp.diff could be used to check the content of the request

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:192]: Following tasks will be done:

- [ ] Factorize code with default case

File: src/control/lib/control/pool.go
Comment (tanabarr) [reply src/control/lib/control/pool.go:999]: You can create a new ticket/issue in JIRA, link it to "this" issue (DAOS-9196) and then reference the new ticket number in this PR to indicate you will resolve the issue in that ticket's PR. The new ticket will probably be groomed by project management and fields updated.

File: src/control/lib/control/pool.go
Comment (tanabarr) [reply src/control/lib/control/pool.go]: metada -> metadata

Comment (knard-intel) [reply src/control/lib/control/pool.go]: > metada -> metadata

Arghhhh, stupid me

File: src/control/lib/control/pool.go
Comment (knard-intel) [reply src/control/lib/control/pool.go:999]: @tanabarr thanks for the method :-)
Following tasks will be done
- [ ] Create JIRA ticket for following up actions
- [ ] Remove useless comments on follow up actions

File: src/control/lib/control/pool.go
Comment (knard-intel) [reply src/control/lib/control/pool.go]: Following tasks will be done:

- [ ] fix spelling errors metada -> metadata

File: src/control/lib/control/pool.go
Comment (kjacque) [reply src/control/lib/control/pool.go:982]: I agree it doesn't hurt anything to check, but with this builtin function, we can assume overflow issues are avoided.

From the [Go specification](https://go.dev/ref/spec#Length_and_capacity):

"At any time the following relationship holds:

0 <= len(s) <= cap(s)
The length of a nil slice, map or channel is 0. The capacity of a nil slice or channel is 0."

IMO we should trust the language spec. We can check for 0 if it still makes sense to do so, but checking for < 0 doesn't really make sense in this case.

Comment (knard-intel) [reply src/control/lib/control/pool.go:982]: Following tasks will be done:

- [ ] Replace confusing test

File: src/control/cmd/dmg/pool.go
Comment (mjmac) [reply src/control/cmd/dmg/pool.go]: Sorry for the churn on this -- it was my intent to suggest that we probably don't need the new "all" flag if we allow the user to specify a percentage to the --size parameter. I don't think it's ideal to have both --size 100% and --all, as it increases the surface that needs to be documented, tested, and maintained.

Let's discuss this during the WG today. I am not sure if @knard-intel is invited to that -- will get him added.

Comment (knard-intel) [reply src/control/cmd/dmg/pool.go]: Following tasks will be done:

- [ ] Remove all option and allow percentage with size option

**Comment by johannlombardi** [comment]
This needs to merge master.

**Comment by knard-intel** [comment]
> This needs to merge master.

done

File: src/control/cmd/dmg/pool.go
Comment (tanabarr) [inline src/control/cmd/dmg/pool.go:71]: does this really require a regex? if so then could we have a comment to make it clear why and what it is for

File: src/control/lib/control/pool_test.go
Comment (tanabarr) [inline src/control/lib/control/pool_test.go:1338]: we normally test for errors and expected operation in the same table-based test

**Review by tanabarr — DISMISSED** [review]
nothing blocking but I think the dmg tests could be tidied up a bit

File: src/control/cmd/dmg/pool.go
Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:71]: Following tasks will be done:
- [ ] Add meaningful comment

From my side, I would like to keep a regex as it allows to clearly define the supported syntax.

File: src/control/lib/control/pool_test.go
Comment (knard-intel) [reply src/control/lib/control/pool_test.go:1338]: With latest, is probably easier to factorize the code.

Following tasks will be done:
- [ ] Factorize test code

**Review by tanabarr — APPROVED** [review]

**Review by mjmac — APPROVED** [review]
Looks good to me. I checked out the PR and played around with it a bit. I think this will improve the user experience. Nice work.

File: src/control/cmd/dmg/pool.go
Comment (kjacque) [inline src/control/cmd/dmg/pool.go:121]: There's no longer an "--all" flag so the error message should be fixed.

File: src/control/cmd/dmg/pool.go
Comment (kjacque) [inline src/control/cmd/dmg/pool.go:129]: Same note, please update the error.

**Review by kjacque — CHANGES_REQUESTED** [review]
Overall looks good. I noticed the error returned to the user still references the --all flag, which ought to be fixed now that it's been factored out.

File: src/control/cmd/dmg/pool.go
Comment (johannlombardi) [reply src/control/cmd/dmg/pool.go:129]: oops, sorry, i saw your comment after landing it. @knard-intel, please address in a new PR. Sorry @kjacque

Comment (kjacque) [reply src/control/cmd/dmg/pool.go:129]: Whoops.

Comment (knard-intel) [reply src/control/cmd/dmg/pool.go:129]: fixed with PR https://github.com/daos-stack/daos/pull/7903

**Pull request merged**
